File: app/src/main/java/com/mohsenoid/rickandmorty/util/config/ConfigProviderImpl.java
package com.mohsenoid.rickandmorty.util.config;

import android.app.Application;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import androidx.annotation.VisibleForTesting;

public class ConfigProviderImpl implements ConfigProvider {

    @VisibleForTesting
    public static ConfigProviderImpl instance;

    private final Application context;

    private ConfigProviderImpl(Application context) {
        this.context = context;
    }

    public static synchronized ConfigProviderImpl getInstance(Application context) {
        if (instance == null)
            instance = new ConfigProviderImpl(context);

        return instance;
    }

    @Override
    public boolean isOnline() {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivityManager == null) return false;

        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        if (activeNetworkInfo == null) return false;

        return connectivityManager.getActiveNetworkInfo().isConnected();
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/util/config/ConfigProviderImpl.kt
package com.mohsenoid.rickandmorty.util.config

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities

/**
 *
 * This class implements [ConfigProvider] and provides configurations required on runtime.
 *
 * @constructor Creates a ConfigProvider using the Android Application context.
 * @param context the Android Application context.
 */
class ConfigProviderImpl(private val context: Context) : ConfigProvider {

    /**
     * This function uses [ConnectivityManager]  to check active Network Capabilities to know
     * the Android phone network connectivity status.
     *
     * @return is the phone connected to a network through cellular or WiFi.
     */
    override fun isOnline(): Boolean {
        val connectivityManager: ConnectivityManager =
            context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
                ?: return false

        val activeNetwork: Network = connectivityManager.activeNetwork ?: return false

        val networkCapabilities: NetworkCapabilities =
            connectivityManager.getNetworkCapabilities(activeNetwork) ?: return false

        return networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
            networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
            networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
    }
}
